flac/decoder.go

```diff
@@ -0,0 +1,109 @@
+// TODO: add seek support; only enabled in the dev branch of mewkiz/flac.
+
+package flac
+
+import (
+	"io"
+
+	"github.com/mewkiz/flac"
+	"github.com/mewkiz/flac/frame"
+	"zikichombo.org/sound"
+	"zikichombo.org/sound/freq"
+)
+
+// Decoder encapsulates state for decoding and seeking a FLAC audio stream.
+type Decoder struct {
+	stream *flac.Stream
+	frame  *frame.Frame // current frame.
+	i      int          // index of current sample in subframe(s).
+}
+
+// NewDecoder creates a decoder from a FLAC audio stream (seekable, readable).
+func NewDecoder(r io.Reader) (*Decoder, error) {
+	stream, err := flac.New(r)
+	if err != nil {
+		return nil, err
+	}
+	d := &Decoder{
+		stream: stream,
+	}
+	return d, nil
+}
+
+func (d *Decoder) Receive(dst []float64) (int, error) {
+	if len(dst)%d.Channels() != 0 {
+		return 0, sound.ErrChannelAlignment
+	}
+	n := 0                       // number of frames (samples per channel) buffered.
+	m := len(dst) / d.Channels() // number of frames (samples per channel) to buffer.
+	bps := int(d.stream.Info.BitsPerSample)
+	for n < m {
+		if d.frame == nil {
+			frame, err := d.stream.ParseNext()
+			if err != nil {
+				return n, err
+			}
+			d.frame = frame
+			d.i = 0
+		}
+		samplesLeft := len(d.frame.Subframes[0].Samples[d.i:])
+		j := m
+		if j > samplesLeft {
+			j = samplesLeft
+		}
+		if n+j > m {
+			j = m - n
+		}
+		if j == 0 {
+			panic("cannot retrieve zero samples")
+		}
+		for c := 0; c < d.Channels(); c++ {
+			toFloats(dst[c*m+n:c*m+n+j], d.frame.Subframes[c].Samples[d.i:d.i+j], bps)
+		}
+		d.i += j
+		n += j
+		if len(d.frame.Subframes[0].Samples[d.i:]) == 0 {
+			d.frame = nil
+			d.i = 0
+		}
+	}
+	if n < m {
+		// TODO: handle n < m. Move frames in dst to their correct position
```

Comment:
How did you handle this case in the WAV decoder? Did you need to perform an extra `copy` to move the samples into the right place? I.e. `dst[c*n:(c+1)*n]` instead of `dst[c*m:(c+1)*m]`.

Reply:
To fit the Receive interface, a copy needs to be done at EOF in case the whole buffer is not filled.  This makes calling Receive vastly easier.  I don't remember what I did in wav offhand, but there is a function sound/cil/Compact you can call which does the copy if necessary, so you can just call cil.Compact(dst, n, m) without even guarding with an if.

BTW, this is actually unreachable (as the only exit from "for n<m" is a return.  The call to cil.Compact could go at line 44 (the return).

Reply:
> To fit the Receive interface, a copy needs to be done at EOF in case the whole buffer is not filled and there is more than 1 channel.

Only at EOF or in case of other errors as well, where partial writes to dst has been made already?

Reply:
> cil.Compact(dst, n, m)

Having a closer look, it should be `cil.Compact(dst, nC, n)`, right? Notice, the `n` rather than `nF` in the last argument. `n` being number of frames read, `nF` being number of frames possible to fit into `dst` (per channel).

Reply:
> BTW, this is actually unreachable (as the only alternate exit from "for n<m" is a return. The call to cil.Compact could go at line 44 (the return).

Another good catch, I've moved this logic now to line 44.

Reply:
The interface for Receive states that partial filling of dst can only happen if subsequently it's EOF.  Other errors I would guess return 0.  The idea is to simplify the caller case handling so there is no need to process things unless it works at expected.


zc/dsp/resample uses cil.Compact.  There is the same functionality in sound.Pipe() but it was implemented before cil.Compact and would require importing sound/cil into sound (in sound/pipe.go#compact).

Reply:
Re: cil.Compact signature, yes you are right, apologies for my earlier imprecision about it.  I am unfortunately susceptible to that sort of thing when going from memory.

BTW: how do you quote previous parts of the conversation in this github interface? I can't figure it out...